### PR TITLE
[move] Remove `Compiled{Script,Module}::freeze`

### PR DESCRIPTION
**Pull request by modocache**
Delete `CompiledScript::freeze` and `CompiledModule::freeze` in favor of explicitly bounds-checking scripts and modules, as a follow-up on #8667.

### CODE REVIEW

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/constants_tests.rs
Comment (tnowacki) [inline language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/constants_tests.rs:43]: You could probably remove these checks if you wanted to

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/duplication_tests.rs
Comment (tnowacki) [inline language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/duplication_tests.rs:17]: Same as the other, you could probably remove these if you wanted

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/generic_ops_tests.rs
Comment (tnowacki) [inline language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/generic_ops_tests.rs:197]: Same as the other, you could probably remove these if you wanted

File: language/bytecode-verifier/src/verifier.rs
Comment (tnowacki) [inline language/bytecode-verifier/src/verifier.rs:28]: Can you comment here? It is actually unsafe to lookup the self_id for the location and would be good to document that

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/multi_pass_tests.rs
Comment (tnowacki) [inline language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/multi_pass_tests.rs:14]: This whole section feels weird. maybe just switch it to calling the main `verify_module`?

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/signature_tests.rs
Comment (tnowacki) [inline language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/signature_tests.rs:19]: Same as the other, you could probably remove these if you wanted

File: language/compiler/ir-to-bytecode/src/compiler.rs
Comment (tnowacki) [inline language/compiler/ir-to-bytecode/src/compiler.rs:480]: Seems sane to keep this here, since the IR is not supposed to be able to produce modules that don't bounds check

File: language/diem-tools/writeset-transaction-generator/src/release_flow/mod.rs
Comment (tnowacki) [inline language/diem-tools/writeset-transaction-generator/src/release_flow/mod.rs:36]: nit I'd remove this. It should be verified elsewhere. If not, we should do that

File: language/diem-tools/writeset-transaction-generator/src/release_flow/unit_tests/release_tests.rs
Comment (tnowacki) [inline language/diem-tools/writeset-transaction-generator/src/release_flow/unit_tests/release_tests.rs:32]: Same as the other writeset check, it should just verify completely

File: language/move-binary-format/src/deserializer.rs
Comment (tnowacki) [inline language/move-binary-format/src/deserializer.rs:14]: Probably a good idea to keep this here, even though it is in the other verifier pass. But we should make sure we aren't duplicating work

Comment (tnowacki) [reply language/move-binary-format/src/deserializer.rs:14]: Seeing this kind of makes me want a wrapper around `CompiledModule` that has absolutely zero functionality just something like `UnverifiedModule(CompiledModule)` with `fn into_inner_I_promise_I_know_what_I_am_doing(self) -> CompiledModule { self.0 }`

This is really what CompiledModuleMut should have been if we wanted that sort of type guarantee. Then `verify_module` can strip this thing out and verify it and return the other module... maybe... might not be possible as some spots only verify by reference...

File: language/move-binary-format/src/file_format.rs
Comment (tnowacki) [inline language/move-binary-format/src/file_format.rs:1979]: What is this function still used for? seems weird and bad

File: language/move-binary-format/src/proptest_types.rs
Comment (tnowacki) [inline language/move-binary-format/src/proptest_types.rs:355]: Same as the other, you could probably remove these if you wanted. Should be verifying elsewhere

File: language/testing-infra/test-generation/src/abstract_state.rs
Comment (tnowacki) [inline language/testing-infra/test-generation/src/abstract_state.rs:482]: Same as the other, you could probably remove these if you wanted

File: language/move-vm/runtime/src/unit_tests/vm_arguments_tests.rs
Comment (tnowacki) [inline language/move-vm/runtime/src/unit_tests/vm_arguments_tests.rs:748]: Same as the other, you could probably remove these if you wanted

File: language/move-vm/integration-tests/src/tests/bad_storage_tests.rs
Comment (tnowacki) [inline language/move-vm/integration-tests/src/tests/bad_storage_tests.rs:249]: Same as the other, you could probably remove these if you wanted

File: language/move-model/src/lib.rs
Comment (tnowacki) [inline language/move-model/src/lib.rs:399]: I don't think this is needed. But could defer to @wrwg. Wolfgang, do want these modules verified?

**Review by tnowacki — COMMENTED** [review]
Looks good! 
- Most of the usecases in tests can probably be removed
- We should look deeper at this interaction with deserializing and verify_module. Might be neat to explore that simple wrapper I mentioned

**Comment by modocache** [comment]
(Responded to and resolved a bunch of review comments, to hopefully make the remaining unresolved ones more prominently displayed in the GitHub UI.)

File: language/move-binary-format/src/file_format.rs
Comment (modocache) [reply language/move-binary-format/src/file_format.rs:1979]: This function is only used by bytecode-verifier-tests, so #8735 might be an improvement: it moves the function, which seems like a convenience helper for tests, out of this core module and closer to the tests that use it, preventing the function from being used more widely.

**Review by tnowacki — APPROVED** [review]
LGTM!

**Comment by tnowacki** [comment]
Looks like there is some proptest issue though? I would definitely verify that before landing

File: language/diem-tools/writeset-transaction-generator/src/release_flow/unit_tests/release_tests.rs
Comment (modocache) [reply language/diem-tools/writeset-transaction-generator/src/release_flow/unit_tests/release_tests.rs]: Just wanted to point one thing out: prior to this pull request and your comment here, `empty_module` would return a module that would fail full verification, due to an `EMPTY_CODE_UNIT` error. In response to your suggestion here, I updated `empty_module` to generate a single `Ret` instruction in its code unit, which allows it to verify.

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/multi_pass_tests.rs
Comment (modocache) [reply language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/multi_pass_tests.rs]: Unfortunately I wasn't able to accomplish this easily. `CompiledModule::valid_strategy` is capable of generating modules that do not pass `friends::verify_module` verification, due to the strategy generating friend declarations that include the module itself, causing `INVALID_FRIEND_DECL_WITH_SELF` validation errors. I think the best course of action is to modify the module generator to no longer generate modules with invalid friend declarations, but I found it pretty difficult to accomplish that as part of this pull request.

I considered three possible approaches:

1. [The proptest book chapter on "Filtering"](https://altsysrq.github.io/proptest-book/proptest/tutorial/filtering.html) describes "local filtering," where the module generation strategy can filter out values with invalid friend declarations. However, while I found it simple to filter out modules with a self index of a specific value, or modules that contained a friend declaration index of a specific value, I couldn't figure out how to tie the two together: filtering out module values where the friend declaration indices included the value that the strategy chose for the self index.
2. That sounds like what [the proptest book refers to as "higher-order strategies,"](https://altsysrq.github.io/proptest-book/proptest/tutorial/higher-order.html) but I found `prop_flat_map` difficult to use, and even after several attempts, I wasn't able to filter out the invalid values. I can go into more detail on what I tried but I think it might be simpler to submit that work as a separate pull request.
3. The proptest book chapter on filtering also mentions "global filtering," in which this test could return `Err(TestCaseError::Reject)` when a module's friend declarations include itself. But that wouldn't fix the module generation strategy, and would instead just get this one test passing.

My preference would be for me to follow up on this by submitting a separate pull request to update module generation with a higher-order strategy that generated friend declarations that excluded the self index.

**Comment by modocache** [comment]
I need another review here: check out the updates to the two conversations I un-resolved, on module generation and on empty code units.

**Review by tnowacki — APPROVED** [review]

**Pull request merged**
